h5py/h5t.pyx

```diff
@@ -222,10 +235,23 @@ cdef dict _order_map = { H5T_ORDER_NONE: '|', H5T_ORDER_LE: '<', H5T_ORDER_BE: '
 cdef dict _sign_map  = { H5T_SGN_NONE: 'u', H5T_SGN_2: 'i' }
 
 # Available floating point types
-available_ftypes = dict()
-for ftype in np.typeDict.values():
-    if np.issubdtype(ftype, float):
-        available_ftypes[np.dtype(ftype).itemsize] = np.finfo(ftype)
+def _get_available_ftypes():
+    def cmp_ftype(t):
+        return np.finfo(t).maxexp
+
+    available_ftypes = defaultdict(list)
+    for ftype in np.typeDict.values():
+        if np.issubdtype(ftype, np.floating):
+            available_ftypes[np.dtype(ftype).itemsize].append(ftype)
+
+    sorted_ftypes = []
+    for size, ftypes in sorted(available_ftypes.items()):
+        for ftype in sorted(ftypes, key=cmp_ftype):
+            if ftype not in sorted_ftypes:
+                sorted_ftypes.append((ftype, np.finfo(ftype), size))
+    return tuple(sorted_ftypes)
+
+available_ftypes = _get_available_ftypes()
```

Comment:
This is an API change, `h5py.h5t.available_ftypes` is accessible to users which means someone is depending on it.  I think the right thing to do here is to leave this dict as-is and create a new private attribute for this data structure.

Reply:
I've replaced `available_ftypes` with `_available_ftypes`, so it's explicit that the variable is internal, and to avoid shadowing the old dict.

The problem with keeping the old dict is that it's not just read only, so users may expect to be able to add additional types to it and have them be supported by h5py (which won't be true). We could make `available_ftypes` a read-only dict which warns when accessed, so that users know not to use `available_ftypes`, and that they can't modify it, but as users tend to ignore warnings, and can still break user code, I'm not sure whether that's an improvement.